status_manager.go

```diff
@@ -0,0 +1,425 @@
+		return fmt.Errorf("cannot start: current status is %s, expected Initializing", currentStatus)
+	}
+
+	// Fetch names in registration order for deterministic startup
+	componentNames := sm.getAllComponents()
+
+	// Track started components so we can roll back on failure
+	startedComponents := make([]string, 0, len(componentNames))
+
+	// Start components in registration order
+	for _, name := range componentNames {
+		wrapper, err := sm.getComponentWrapper(name)
+		if err != nil {
+			sm.rollbackStartup(startedComponents)
+			return err
+		}
+
+		component := wrapper.GetComponent()
+
+		wrapper.SetStatus(ComponentStatusStarting)
+
+		// Start the component to transition it to running
+		if err := component.Start(sm.ctx); err != nil {
+			wrapper.SetStatus(ComponentStatusFailed)
+
+			// Report the startup error via the handler
+			wrappedErr := sm.errorHandler.HandleStartupError(name, err)
+
+			// Roll back the components that already started
+			sm.rollbackStartup(startedComponents)
+
+			return wrappedErr
+		}
+
+		// Mark the wrapper as running
+		wrapper.SetStatus(ComponentStatusRunning)
+		startedComponents = append(startedComponents, name)
+	}
+
+	// All components started, transition to Open
+	return sm.transitionTo(StatusOpen)
+}
+
+// rollbackStartup stops started components in reverse order to undo partial start-ups
+func (sm *StatusManager) rollbackStartup(startedComponents []string) {
+	// Stop components in reverse order of startup
+	for i := len(startedComponents) - 1; i >= 0; i-- {
+		name := startedComponents[i]
+		wrapper, err := sm.getComponentWrapper(name)
+		if err != nil {
+			continue
+		}
+
+		component := wrapper.GetComponent()
+		wrapper.SetStatus(ComponentStatusStopping)
+
+		// Use a shorter timeout for the rollback stop
+		if err := component.Stop(5 * time.Second); err != nil {
+			wrapper.SetStatus(ComponentStatusFailed)
+		} else {
+			wrapper.SetStatus(ComponentStatusStopped)
+		}
+	}
+}
+
+// Close terminates all components in reverse registration order for graceful shutdown
+func (sm *StatusManager) Close() error {
+	// Use mutex to prevent concurrent Close() calls from racing
+	sm.componentsMu.Lock()
+	currentStatus := sm.Status()
+
+	// Idempotent: if already Closed, return immediately
+	if currentStatus == StatusClosed {
+		sm.componentsMu.Unlock()
+		return nil
+	}
+
+	// If already closing, wait for completion
+	if currentStatus == StatusClosing {
+		sm.componentsMu.Unlock()
+		return sm.waitForClosed()
+	}
+
+	// Transition to Closing state while holding the lock
+	if err := sm.transitionTo(StatusClosing); err != nil {
+		sm.componentsMu.Unlock()
+		return err
+	}
+	sm.componentsMu.Unlock()
+
+	// Cancel the shared context to signal components to stop
+	sm.cancel()
+
+	// Collect component names and reverse them so shutdown happens in reverse registration order
+	componentNames := sm.getAllComponents()
+
+	// Reverse the list to compute shutdown order
+	shutdownOrder := make([]string, len(componentNames))
```

Comment:
no need to create a new slices, just let `shutdownOrder := sm.getAllComponents()` then reverse it is better.